pkg/cvo/upgradeable.go

```diff
@@ -234,6 +247,131 @@ func (check *clusterManifestDeleteInProgressUpgradeable) Check() *configv1.Clust
+	cvLister         configlistersv1.ClusterVersionLister
+	cvoName          string
+}
+
+func (check *clusterAdminAcksCompletedUpgradeable) Check() *configv1.ClusterOperatorStatusCondition {
+	cv, err := check.cvLister.Get(check.cvoName)
+	if meta.IsNoMatchError(err) || apierrors.IsNotFound(err) {
+		message := fmt.Sprintf("Unable to get ClusterVersion, err=%v.", err)
+		klog.Error(message)
+		return &configv1.ClusterOperatorStatusCondition{
+			Type:    configv1.ClusterStatusConditionType("UpgradeableAdminAckRequired"),
+			Status:  configv1.ConditionFalse,
+			Reason:  "UnableToGetClusterVersion",
+			Message: message,
+		}
+	}
+	currentVersion := clusterversion.GetCurrentVersion(cv.Status.History)
+
+	// This can occur in early start up when the configmap is first added and version history
+	// has not yet been populated.
+	if currentVersion == "" {
+		return nil
+	}
+
+	var gateCm *corev1.ConfigMap
+	if gateCm, err = check.adminGatesLister.Get(internal.AdminGatesConfigMap); err != nil {
+		var message string
+		if apierrors.IsNotFound(err) {
+			message = fmt.Sprintf("%s configmap not found.", internal.AdminGatesConfigMap)
+		} else if err != nil {
+			message = fmt.Sprintf("Unable to access configmap %s, err=%v.", internal.AdminGatesConfigMap, err)
+		}
+		klog.Error(message)
+		return &configv1.ClusterOperatorStatusCondition{
+			Type:    configv1.ClusterStatusConditionType("UpgradeableAdminAckRequired"),
+			Status:  configv1.ConditionFalse,
+			Reason:  "UnableToAccessAdminGatesConfigMap",
+			Message: message,
+		}
+	}
+	var ackCm *corev1.ConfigMap
+	if ackCm, err = check.adminAcksLister.Get(internal.AdminAcksConfigMap); err != nil {
+		var message string
+		if apierrors.IsNotFound(err) {
+			message = fmt.Sprintf("%s configmap not found.", internal.AdminAcksConfigMap)
+		} else if err != nil {
+			message = fmt.Sprintf("Unable to access configmap %s, err=%v.", internal.AdminAcksConfigMap, err)
+		}
+		klog.Error(message)
+		return &configv1.ClusterOperatorStatusCondition{
+			Type:    configv1.ClusterStatusConditionType("UpgradeableAdminAckRequired"),
+			Status:  configv1.ConditionFalse,
+			Reason:  "UnableToAccessAdminAcksConfigMap",
+			Message: message,
+		}
+	}
+	var reasons []string
+	var messages []string
+	for k, v := range gateCm.Data {
+		if reason, message := checkAdminGate(k, v, currentVersion, ackCm); reason != "" {
+			reasons = append(reasons, reason)
+			messages = append(messages, message)
+		}
+	}
+	if len(reasons) == 1 {
+		return &configv1.ClusterOperatorStatusCondition{
+			Type:    configv1.ClusterStatusConditionType("UpgradeableAdminAckRequired"),
+			Status:  configv1.ConditionFalse,
+			Reason:  reasons[0],
+			Message: messages[0],
+		}
+	} else if len(reasons) > 1 {
+		return &configv1.ClusterOperatorStatusCondition{
+			Type:    configv1.ClusterStatusConditionType("UpgradeableAdminAckRequired"),
+			Status:  configv1.ConditionFalse,
+			Reason:  "MultipleReasons",
+			Message: strings.Join(messages, " "),
```

Comment:
`return "AdminAckRequired", gateValue` in the "you haven't ack'ed this gate yet" case could be fairly long, right?  Maybe we want `fmt.Sprintf("Multiple admin acks failed:\n* %s", strings.Join(messages, "\n* "))` here, like we have for [collecting multiple precondition check errors][1]?  Although there's a risk that life gets complicated while some higher level applies its own concatenation logic to glue together a bunch of `Upgradeable=False` reasons...

[1]: https://github.com/openshift/cluster-version-operator/blob/77a24f77b8f6e9975b5f41a787d6368ff609e391/pkg/payload/precondition/precondition.go#L84

Reply:
Collapse down to single Reason with sorted messages.